python/edit_fails_by_row.py

```diff
@@ -0,0 +1,99 @@
+import math
+import json
+import os
+import pandas as pd
+import random
+import string
+import time
+import yaml
+import utils
+
+from lar_generator import lar_gen #Imports lar_gen class. 
+
+from large_test_files import LargeTestFiles
+
+from large_test_files import unique_uli
+
+class EditFailsByRow(object):
+
+	def __init__(self, passes_all_filepath, passes_all_filename, 
+		fails_all_filepath, fails_all_filename):
+		"""The class instantiates with a filepath and name to a file that 
+		passes every edit, and a filepath and name to a file that fails 
+		a an edit or multiple edits for every row."""
+
+		#To create a file that has a certain number of clean rows and failed rows,
+		# two sets of TS and LAR data are loaded to create a file that
+		# has some rows that fail edits, and some rows that do not. 
+		# The constructor provides a way for users to input their files that pass cleanly
+		# and files that fail an edit or multiple edits for every row.  
+		
+		#Loading TS and LAR data for the clean file. 
+		self.clean_ts, self.clean_lar = utils.read_data_file(path=passes_all_filepath, 
+                                                 data_file=passes_all_filename)
+
+		#Loading TS and LAR data for the file that fails a set of edits for every row. 
+		self.fail_ts, self.fail_lar = utils.read_data_file(path=fails_all_filepath,
+												 data_file=fails_all_filename)
+
+		#Storing the first LAR row of the clean file. 
+		self.clean_lar = self.clean_lar.iloc[0:1]
+		#print(self.clean_lar)
+
+		#Storing the LEI in a global variable. 
+		self.lei = self.clean_ts.iloc[0][14]
+
+		#Storing the bank name as a global variable. 
+		self.bank_name = self.clean_ts.iloc[0][1]
+
+		#Setting LEI's for each by self.lei.
+		self.clean_lar["lei"] = self.lei
+
+		self.fail_lar["lei"] = self.lei
+
+		#Storing the first LAR row of the file that fails a set of edits for every row. 	
+		self.fail_lar = self.fail_lar.iloc[0:1]
+		#print(self.fail_lar)
+
+		#Prints a statement indicating that the object has been instantiated. 
+		print("Instantiated EditFailsbyRow Class")
+
+	def create_edit_fails_by_row_file(self, rows_failed=50, rows_total=100, 
+		output_filepath= "../edits_files/edit_fails_by_row/"):
+		"""Creates a file that fails a set of edits by a 
+		certain number of rows."""
+
+		if rows_failed >= rows_total:
+			print("Sorry - rows_failed must be a number less than rows_total.")
```

Comment:
What do you think about setting rows_failed = rows_total and printing a message that says what was done and why? 

I think this boils down to whether we want to have a hard stop, or provide a way to continue without changing input parameters. What do you think?

Reply:
Because the code already outputs a print statement of what is done, I agree, it would be useful to build in ways to allow the code to make an output without breaking. 

One issue might be that the final file is made through concatenating passing LAR rows and failing LAR rows. The number of rows passed is included in the file by concatenating the passing LAR rows by a multiplier of the remaining rows (total rows - failed rows). 

Setting rows_failed to rows_total would set this multiplier to 0, and it would not concatenate. 

We can change how the files are put together, or we can evaluate so that any value for rows_failed higher than it should be would be evaluated to one less than the rows_total.

Reply:
I see your point. Let's leave this as it is and post an issue on the refactor project if we want to address it later.